app/services/promo_registrar.rb

```diff
@@ -12,44 +15,47 @@ def perform
     channels.each do |channel|
       if should_register_channel?(channel)
         referral_code = register_channel(channel)
-        promo_registration = PromoRegistration.new(channel_id: channel.id, promo_id: active_promo_id, referral_code: referral_code)
+        promo_registration = PromoRegistration.new(channel_id: channel.id, promo_id: @promo_id, referral_code: referral_code)
```

Comment:
minor: to increase code readability, consider checking for referral code and short circuiting:
```ruby
  next if referral_code.nil?
```

this also protects from the `#save!` exception case below